rfc/mediachain-rfc-1.md

```diff
@@ -0,0 +1,435 @@
+
+Status: DRAFT
+
+Author: vyzo
+
+## Overview
+
+The Mediachain[1] is a decentralized system designed for tracking
+metadata related to Creative Works and Internet Media.
+It is being developed as an Open Source, Open Data system with
+decentralized stakeholders.
+
+This document serves as a  specification of the Mediachain data
+model and the shape of the distributed store at the heart of
+system.
+It is intended to provide a stable basis for bootstrapping the
+system from a small-scale proof of concept to a fully distributed,
+Open Data Peer-to-Peer network.
+
+## The Mediachain Data Model
+
+### Data Objects
+
+The Mediachain is stored in a distributed append-only datastore with
+content addressable access. This maps well to the IPFS[2] peer-to-peer
+network, which names and accesses data using the SHA256 hash of
+the content. Objects stored in the network are immutable and persistent;
+their hashes become their _Canonical_ identifiers through which they can
+be referenced.
+
+Mediachain objects follow the IPLD[3] data model,
+allowing for storage of key-valued pairs where keys
+are strings and values are other key-valued pairs or strings.
+
+Thus we have:
+```
+DataObject = {
+ "type" : <String>
+ <Key> : <Value>  ...
+ }
+
+Key   = <String>
+Value = <String> | <CompoundValue>
+CompoundValue = {
+ <Key> : <Value> ...
+ }
+
+```
+All Mediachain data objects have a mandatory `type` field that represents
+the data type of the object.
+Links can be embedded between data objects by using the IPLD `@link`
+special key, allowing us to construct Merkle DAGs.
+
+### Entities and Artefacts
+
+There are two principal types of objects in the mediachain:
+entities and artefacts. An _Entity_ represents a person or organization
+who may associate with artefacts, for example by creating or posting
+a reference to the artefact. _Artefacts_ represent media works as tracked
+by the Mediachain. They can be images, video, text, or any other
+common media found in the Internet. 
+
+In the Mediachain schema, entities are instances of data objects with
+their type field set to `entity` and a mandatory `name` field.
+Artefacts are instances of data objects with their type field set
+to `artefact`, a mandatory name and optional `description` and creation
+date fields.
+
+Artefacts can also have associated their data stored in IPFS, so that media
+can be directly accessed from references to their Canonicals.
+If this is the case, then the artefact object will contain a link to
+the IPFS datablob in its `data` field
+
+Finally, both types carry a set of cryptographic signatures that assert their
+validity. The signatures come from _signatories_ with known keys in the
+system; any entity that contains a public key in its data can act
+as a signatory.
+
+We can make things more concrete with the following schema:
+```
+Entity = {
+ "type" : "entity"
+ "name" : <String>
+ ["key" : <Reference>]
+ "signatures" : <Signatures>
+ <Key> : <Value> ... ; entity metadata
```

Comment:
should metadata be held in a compound value? this helps us prevent problems with duplicate keys, i.e. `name` being a piece of the metadata for some future object.

```
"metadata": <CompoundValue>
```